### PR TITLE
Add failing SAF test that shows problem with bigger tbl file

### PR DESCRIPTION
**Pull request by Apanatshka**
@mpsijm Here's the problem in the form of a JUnit test.

### CODE REVIEW

**Comment by Apanatshka** [comment]
I think I found and fixed the problem. `StrategoArrayList` was being used and skipped the first element in a list during `doSlowMatch` (a bug). But this usually isn't triggered because the `hashCode` is checked first. Therefore only two lists in the `SAFWriter` with hashcollisions were replaced with the wrong other list that was already in the cache (used for maximal sharing in the binary format).

File: org.spoofax.terms/test/org/spoofax/interpreter/terms/IStrategoListTests.java [hunk 1]
```diff
@@ -53,6 +53,25 @@ default IStrategoList createIStrategoList(@Nullable List<IStrategoTerm> elements
                     elements.size()), null, null), annotations, attachments);
         }
 
+        /**
+         * Creates a new instance of {@link IStrategoList} for testing.
```
[the remark this answers is not in the document]
Comment (Apanatshka) [reply org.spoofax.terms/test/org/spoofax/interpreter/terms/IStrategoListTests.java]: I added versions called `WithHashCode` that have a proper hashcode implementation for the tests that check that the hashCode implementations of all `IStrategoList` give the same hashcode. I see now that that's a confusing name, so I'll edit the names. And I'll add some text in the different JavaDocs that typical `createIStratego*` methods have a fixed hashcode of `0`.

[the remark this answers is not in the document]
Comment (Apanatshka) [reply org.spoofax.terms/test/org/spoofax/interpreter/terms/IStrategoListTests.java]: @Virtlink does that resolve this comment?

[the remark this answers is not in the document]
Comment (Virtlink) [reply org.spoofax.terms/test/org/spoofax/interpreter/terms/IStrategoListTests.java]: Yes, looks good.

**Pull request merged**